emr/launch_emr.py

```diff
@@ -8,75 +9,161 @@
                     {
-                        "Market": "ON_DEMAND",
+                        "Market": cfg.get("master_instance", {}).get(
+                            "market", "ON_DEMAND"
+                        ),
                         "InstanceRole": "MASTER",
                         "InstanceType": cfg.get("master_instance", {}).get(
-                            "type"
+                            "type", "m5.2xlarge"
                         ),
                         "InstanceCount": 1,
                     },
                     {
-                        "Market": "ON_DEMAND",
+                        "Market": cfg.get("core_instance", {}).get(
+                            "market", "ON_DEMAND"
+                        ),
                         "InstanceRole": "CORE",
                         "InstanceType": cfg.get("core_instances", {}).get(
-                            "type"
-                        ),
-                        "InstanceCount": cfg.get("core_instances", {}).get(
-                            "count"
+                            "type", "r5.12xlarge"
                         ),
+                        "InstanceCount": cfg.get("core_instances", {}).get("count", 2),
                     },
                 ],
-                "Ec2KeyName": cfg["ssh_keys"]["key_name"],
+                "Ec2KeyName": key_name,
                 "KeepJobFlowAliveWhenNoSteps": True,
                 "Ec2SubnetId": cfg.get("subnet_id"),
+                "TerminationProtected": False,
+                # "AutoTerminate": True,
             },
-            Applications=[{"Name": "Spark"}],
+            Configurations=[
+                {
+                    "Classification": "yarn-site",
+                    "Properties": {
+                        "yarn.log-aggregation-enable": "true",
+                        "yarn.log-aggregation.retain-seconds": "-1",
+                        "yarn.nodemanager.remote-app-log-dir": "s3://nrel-dsgrid-int-scratch/logs",
+                    },
+                }
+            ],
+            Applications=[
+                {
+                    "Name": "Hadoop",
+                },
+                {
+                    "Name": "Spark",
+                },
+                {
+                    "Name": "Livy",
+                },
+                {
+                    "Name": "Hive",
+                },
+                {
+                    "Name": "JupyterEnterpriseGateway",
+                },
+            ],
             BootstrapActions=[
                 {
                     "Name": "launchFromS3",
                     "ScriptBootstrapAction": {
                         "Path": bootstrap_script_loc,
-                        "Args": [],
+                        "Args": ["--s3scratch", s3_scratch],
                     },
                 },
             ],
             VisibleToAllUsers=True,
+            AutoTerminationPolicy={
+                "IdleTimeout": 3600,  # sec
+            },
```

Comment:
This is an example of an option a user might want to customize/pass in an option for. (Although as long as this is pointing to the dsgrid cloud account, I'd want this script to enforce a maximum idletimeout length.)